### ISSUE TITLE
Bug? View of child controller not recreated when parent controller has RETAIN_DETACH

### ISSUE THREAD

**Issue by stephanschuster** [bluelinelabs/Conductor#165]
I have the following controller/backstack hierarchy (simplified):

```
FrameController
    HomeController
        ImageTileController2
        ImageTileController1
    MapsController
```

I think it doesn't matter but anyway: FrameController extends "SwitchController" as described in [Issue #155](https://github.com/bluelinelabs/Conductor/issues/156). This simply means, that switching between HomeController and MapsController is done by reordering the backstack via `Router.setBackstack([Home, Maps] or [Maps, Home])`.

**The problem:** If HomeController is set with RetainViewMode.RETAIN_DETACH, switching back from Maps to Home does not recreate the view of "ImageTileController2". HomeController.attach() is called, but nothing more (e.g. ImageTileController2.onCreateView()).

Maybe the problem gets clearer with some debug output of my application.

```
Legend:

IID = c.getInstanceId() (abbreviated)
R = c.getRetainViewMode() == Controller.RetainViewMode.RETAIN_DETACH
V = c.getView() != null
A = c.isAttached()
D = c.isDestroyed() || c.isBeingDestroyed()
B = c.getArgs().size()

"x" = true
"_" = false
```

**Use case 1**: HomeController has **RELEASE_DETACH** --> works but is not what I want

```
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[634] _ x x _ 0    FrameController
[f1a] _ x x _ 0        HomeController
[f8c] _ x x _ 1            ImageTileController
[32d] _ _ _ _ 1            ImageTileController
[a2e] _ x _ _ 0        MapsController
------------------------------------------------------------
// Switching from Home to Maps
[a2e] MapsController.onAttach()
[f1a] HomeController.onDetach()
[f1a] HomeController.onSaveViewState()
[f1a] HomeController.onDestroyView()
[f8c] ImageTileController.onDetach()
[f8c] ImageTileController.onSaveViewState()
[f8c] ImageTileController.onDestroyView()
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[634] _ x x _ 0    FrameController
[a2e] _ x x _ 0        MapsController
[f1a] _ _ _ _ 0        HomeController
[f8c] _ _ _ _ 1            ImageTileController
[32d] _ _ _ _ 1            ImageTileController
------------------------------------------------------------
// Switching back from Maps to Home
[f1a] HomeController.onCreateView()
[f8c] ImageTileController.onCreateView()
[f8c] ImageTileController.onRestoreViewState()
[f1a] HomeController.onRestoreViewState()
[f1a] HomeController.onAttach()
[f8c] ImageTileController.onAttach()
[a2e] MapsController.onDetach()
[a2e] MapsController.onSaveViewState()
[a2e] MapsController.onDestroyView()
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[634] _ x x _ 0    FrameController
[f1a] _ x x _ 0        HomeController
[f8c] _ x x _ 1            ImageTileController         <--- GOOD: view != null, attached == true
[32d] _ _ _ _ 1            ImageTileController
[a2e] _ _ _ _ 0        MapsController
------------------------------------------------------------
```

**Use case 2**: HomeController has **RETAIN_DETACH** --> old view of ImageTileController is still visible when switching back, but it has actually been destroyed (listeners removed etc.) and never been recreated.

```
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[4e5] _ x x _ 0    FrameController
[768] x x x _ 0        HomeController
[c05] _ x x _ 1            ImageTileController
[231] _ _ _ _ 1            ImageTileController
[528] x x _ _ 0        MapsController
------------------------------------------------------------
// Switching from Home to Maps
[528] MapsController.onAttach()
[768] HomeController.onDetach()
[c05] ImageTileController.onDetach()
[c05] ImageTileController.onSaveViewState()
[c05] ImageTileController.onDestroyView()
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[4e5] _ x x _ 0    FrameController
[528] x x x _ 0        MapsController
[768] x x _ _ 0        HomeController
[c05] _ _ _ _ 1            ImageTileController
[231] _ _ _ _ 1            ImageTileController
------------------------------------------------------------
// Switching back from Maps to Home
[768] HomeController.onAttach()
[528] MapsController.onDetach()
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[4e5] _ x x _ 0    FrameController
[768] x x x _ 0        HomeController
[c05] _ _ _ _ 1            ImageTileController         <--- BAD: no view, not attached
[231] _ _ _ _ 1            ImageTileController
[528] x x _ _ 0        MapsController
------------------------------------------------------------
```

My first shot was to override HomeController.onAttach() like this:

```
    @Override
    protected void onAttach(@NonNull View view) {
        logLifecycle(".onAttach()", true);
        super.onAttach(view);

        // TODO: Temporary workaround for issue 165 in conductor
        // Link: https://github.com/bluelinelabs/Conductor/issues/165
        if (getRetainViewMode() == RetainViewMode.RETAIN_DETACH) {
            for (Router childRouter : getChildRouters()) {
                childRouter.rebindIfNeeded();
            }
        }
    }
```

BTW: Controller.detach() also seems to somehow delegate to child routers. Controller.attach() does not.

Output looks pretty good then:

```
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[e90] _ x x _ 0    FrameController
[2c8] x x x _ 0        HomeController
[918] _ x x _ 1            ImageTileController
[8f3] _ _ _ _ 1            ImageTileController
[5c3] x x _ _ 0        MapsController
------------------------------------------------------------
// Switching from Home to Maps
[5c3] MapsController.onAttach()
[2c8] HomeController.onDetach()
[918] ImageTileController.onDetach()
[918] ImageTileController.onSaveViewState()
[918] ImageTileController.onDestroyView()
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[e90] _ x x _ 0    FrameController
[5c3] x x x _ 0        MapsController
[2c8] x x _ _ 0        HomeController
[918] _ _ _ _ 1            ImageTileController
[8f3] _ _ _ _ 1            ImageTileController
------------------------------------------------------------
// Switching back from Maps to Home
[2c8] HomeController.onAttach()
[918] ImageTileController.onCreateView()               <--- GOOD: view is recreated
[918] ImageTileController.onRestoreViewState()
[918] ImageTileController.onAttach()
[5c3] MapsController.onDetach()
------------------------------------------------------------
[IID] R V A D B    Controller (Tag)
------------------------------------------------------------
[e90] _ x x _ 0    FrameController
[2c8] x x x _ 0        HomeController
[918] _ x x _ 1            ImageTileController         <--- GOOD: view != null, attached == true
[8f3] _ _ _ _ 1            ImageTileController
[5c3] x x _ _ 0        MapsController
------------------------------------------------------------
```

My remaining problem: the OLD view of the ImageTileController (the one which was active before switching to Maps) seems to be still visible when coming back to Home. Even tough the old view is below (z-index) the newly created view, it shortly gets visible on animations like flip.

**IS THIS A BUG?
Or am I doing something wrong?
Is there a better solution/workaround than `rebindIfNeeded()`?
How can I get rid of this "old view"?**

Any help appreciated.

Thanks for reading till here!

**Comment by stephanschuster** [comment]
x

**Comment by stephanschuster** [comment]
Regarding the above mentioned issue with the remaining **"old view"**: I think this is a separate problem/bug.

In my opinion the view of a child controller with RELEASE_DETACH (see "ImageTileController" above) should be removed from its view container when the view is destroyed and in case the parent controller has RETAIN_DETACH set (see "HomeController" above). However, this seems not to be the case.

Look again at the two lifecycle outputs above when "Switching from Home to Maps":

**Use case 1**: HomeController has **RELEASE_DETACH**

```
// Switching from Home to Maps
[a2e] MapsController.onAttach()
[f1a] HomeController.onDetach()
[f1a] HomeController.onSaveViewState()
[f1a] HomeController.onDestroyView()
[f8c] ImageTileController.onDetach()
[f8c] ImageTileController.onSaveViewState()
[f8c] ImageTileController.onDestroyView()
```

**Use case 2**: HomeController has **RETAIN_DETACH**

```
// Switching from Home to Maps
[528] MapsController.onAttach()
[768] HomeController.onDetach()
[c05] ImageTileController.onDetach()
[c05] ImageTileController.onSaveViewState()
[c05] ImageTileController.onDestroyView()
```

In both cases ImageTileController.onDestroyView() is called, but the view itself is never removed from the container. In use case 1 this does not matter because the parent view is indeed removed and therefore the child view as well. In use case 2, however, this leeds to an unbound orphan view which still exists in the view container when we later go back from Maps to Home.

My current workaround for this looks like this in my BaseController:

```
    @Override
    protected void onDestroyView(@NonNull View view) {
        logLifecycle(".onDestroyView()", false);
        super.onDestroyView(view);

        // TODO: Temporary workaround for issue 165 in conductor
        // Link: https://github.com/bluelinelabs/Conductor/issues/165
        Controller parentController = getParentController();
        ViewParent routerContainer = view.getParent();
        if (parentController != null &&
                !parentController.isAttached() &&
                parentController.getRetainViewMode() == RetainViewMode.RETAIN_DETACH &&
                routerContainer instanceof ViewGroup) {
            ((ViewGroup) routerContainer).removeView(view);
        }

        mUnbinder.unbind();
        mUnbinder = null;
    }
```

With Workaround 1 (overriding onAttach() in my BaseController like shown in my first posting) and Workaround 2 (overriding onDestroyView() in my BaseController like shown above), my app now works as expected.

But I don't like workarounds ... ;-)

Again, it would be nice to hear the experts on both issues.

Cheers.

**Comment by cerisier** [comment]
I seem to have the same issue. Will try to dig more into it tomorrow and maybe come with a fix.

**Comment by EricKuck** [comment]
Can you try with the latest snapshot build? It _should_ be working correctly now.

**Comment by EricKuck** [comment]
I'm pretty sure this has been fixed in the 2.0.5 release. If you're still seeing issues, feel free to reopen.

**Comment by martinflorek** [comment]
@EricKuck I have just hit this same bug in the 2.1.4 version. Is this a regression or it was not fixed yet? Also the two proposed workarounds fix the behaviour.
